etf-sdk/src/lib.rs

```diff
@@ -1,237 +1,246 @@
-
-//     // #[wasm_bindgen_test]
-//     // pub fn wrapper_can_decrypt() {
-//     //     let message_js = serde_wasm_bindgen::to_value(b"test").unwrap();
-//     //     let slot_ids = vec![vec![1, 2, 3], vec![2, 3, 4]];
-//     //     let slot_ids_js = serde_wasm_bindgen::to_value(&slot_ids).unwrap();
-
-//     //     let s = Fr::rand(&mut test_rng());
-//     //     let g = G2::rand(&mut test_rng());
-//     //     let p: G2 = g.mul(s).into();
-//     //     let g_bytes = convert_to_bytes::<G2, 96>(g).to_vec();
-//     //     let p_bytes = convert_to_bytes::<G2, 96>(p).to_vec();
-//     //     let x1 = serde_wasm_bindgen::to_value(&g_bytes).unwrap();
-//     //     let x2 = serde_wasm_bindgen::to_value(&p_bytes).unwrap();
-//     //     let etf = EtfApiWrapper::create(x1, x2);
-//     //     match etf.encrypt(message_js, slot_ids_js, 3) {
-//     //         Ok(ct) => {
-//     //             let t: crypto::client::client::AesIbeCt = serde_wasm_bindgen::from_value(ct).unwrap();
-//     //             let ct_bytes = serde_wasm_bindgen::to_value(&t.aes_ct.ciphertext).unwrap();
-//     //             let nonce_bytes = serde_wasm_bindgen::to_value(&t.aes_ct.nonce).unwrap();
-//     //             let capsule_bytes = serde_wasm_bindgen::to_value(&t.etf_ct).unwrap();
-//     //             // calc valid secrets d = sQ
-//     //             let secrets: Vec<Vec<u8>> = slot_ids.iter().map(|id| {
-//     //                 let q = hash_to_g1(&id);
-//     //                 let d = q.mul(s);
-//     //                 convert_to_bytes::<G1, 48>(d.into()).to_vec()
-//     //             }).collect::<Vec<_>>();
-
-//     //             let sks = serde_wasm_bindgen::to_value(&secrets).unwrap();
-
-//     //             match etf.decrypt(ct_bytes, nonce_bytes, capsule_bytes, sks) {
-//     //                 Ok(m_js) => {
-//     //                     let m: Vec<u8> = serde_wasm_bindgen::from_value(m_js).unwrap();
-//     //                     assert_eq!(m, b"test".to_vec());
-//     //                 }, 
-//     //                 Err(_) => {
-//     //                     panic!("test should pass, but decryption failed");
-//     //                 }
-//     //             }
-//     //         },
-//     //         Err(_) => {
-//     //             panic!("test should pass, but encryption failed");
-//     //         }
-//     //     }
-//     // }
-// }
\ No newline at end of file
+use ark_serialize::CanonicalSerialize;
+use serde::Deserialize;
+#[cfg_attr(tarpaulin, skip)]
+use wasm_bindgen::prelude::*;
+use etf_crypto_primitives::{self, encryption::tlock::{DecryptionResult, SecretKey, TLECiphertext}, ibe::fullident::{IBESecret, Identity}, utils::{self, *}};
+use w3f_bls::{EngineBLS, TinyBLS377};
+use rand_chacha::ChaCha20Rng;
+use ark_std::rand::SeedableRng;
+use w3f_bls::{
+    DoublePublicKey,
+    DoublePublicKeyScheme,
+};
+use serde::Serialize;
+use serde_big_array::BigArray;
+use ark_serialize::CanonicalDeserialize;
+
+ #[wasm_bindgen]
```

Comment:
Can you add docs?

Reply:
Docs have been added for encrypt function, decrypt function, generate_keys function, extract_signature functions, the logging struct, and the keychain struct